File: 2016/09/09.py
from parse import parse

def parser(data): return data

def solver(file): 
    yield get_file_size(file, len)
    yield get_file_size(file, rec := lambda x: get_file_size(x, rec))


def get_file_size(file, func):
    for i, c in enumerate(file):
        if c == '(':
            width, rep, _ = parse('({:d}x{:d}){}',file[i:]) 
            length = i + len(str(width)+str(rep)) + 3
            return i + get_file_size(file[length + width:], func) + rep * func(file[length: length + width])
    return len(file)

File: 2016/09/09.rs
aoc::main!();

fn preprocessing(input: &str) -> &str {
    input
}

fn part_1(file: &str) -> usize {
    solver(file, |c| c.len())
}

fn part_2(file: &str) -> usize {
    solver(file, part_2)
}

fn solver(file: &str, func : impl Fn(&str) -> usize) -> usize {
    let d = &file.splitn(4, &['(','x',')']).collect_vec();

    match d.len() {
        1 => file.len() ,
        4 => { 
            let (previous, width, rep, next) = (d[0],d[1].parse::<usize>().unwrap(), d[2].parse::<usize>().unwrap(),d[3]) ;
            rep * func(&next[..width]) + previous.len() + solver(&next[width..], func)},
        _ => unreachable!()}
}